File: 438. Find All Anagrams in a String/solution.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
  if (s.length < p.length) return [];
  let freqs = new Array(256).fill(0);
  for (let i = 0; i < p.length; i++) {
    freqs[p.charCodeAt(i)]++;
  }
  let matches = 0;
  for (let i = 0; i < p.length - 1; i++) {
    if (freqs[s.charCodeAt(i)]-- > 0) matches++;
  }
  let res = [];
  for (let i = 0; i < s.length - p.length + 1; i++) {
    if (freqs[s.charCodeAt(i + p.length - 1)]-- > 0) matches++;
    if (matches === p.length) res.push(i);
    if (++freqs[s.charCodeAt(i)] > 0) matches--;
  }
  return res;
};

File: 438. Find All Anagrams in a String/solution.go
func findAnagrams(s string, p string) []int {
	if len(s) < len(p) {
		return make([]int, 0)
	}

	var count [26]int
	matches := 0
	result := make([]int, 0, 5)
	for i := 0; i < len(p); i++ {
		count[p[i]-'a']++
	}

	l := 0
	for r := 0; r < len(s); r++ {
		count[s[r]-'a']--
		if (count[s[r]-'a']) >= 0 {
			matches++
		}

		if r >= len(p) {
			count[s[l]-'a']++
			if (count[s[l]-'a']) > 0 {
				matches--
			}
			l++
		}

		if matches == len(p) {
			result = append(result, l)
		}
	}

	return result
}